### ISSUE TITLE
Tag to include + filter

### ISSUE THREAD

**Issue by mercmobily** [leizongmin/tinyliquid#33]
Hi,

First of all, thank you for your great module!
I am embedding it into Ryver, my static web site generator. Ryver is based on filters: basically, the result of a web page is the result of a bunch of filters.
I need to give the opportunity to include a file, and filter it through one of Ryver's filters (for example, `markup-markdown`.
I can see two path:

# Filters 

I could:

* define a `markup-markdown` filter in tinyliquid (quite trivial)
* Have code like this:
````
    {%capture result%}
      {% include _included.md %}
    {%endcapture%}
    {{ result | ryver-markup-markdown }}
````
However, syntax-wise this is less than ideal.

# Define a tinyliquid  custom tag

I could compile my contents with:

    var render = liquid.compile( fileData.contents, { customTags: { filter_include: function(){ ... } } } );

Here, filter_include would work just like `include`; however, it would have a 'filter' option:

      {% filter_include _included.md filters=frontmatter,markdown %}
   
However, I am finding this one much harder to code; I wanted to reuse the code from [Contents.prototype.include](https://github.com/leizongmin/tinyliquid/blob/master/lib/context.js#L557), but I am having trouble understanding it: the parameters are not `(context, name, body)` and I feel a little lost. Although. this would me my ideal solution.

Thank you in advance!

**Comment by leizongmin** [comment]
We can use `include` tag like this:

```
// filename is a string
{% include xxx %} or {% include "xxx" %}

// filename is a variable
{% include {{xx}} %} and with filters: {% include {{xx | yy}} %}

// assign a variable as the top level locals
{% include xxx with yy %}

// the same as: {% assign a=1 %} {% assign b=2 %} {% include xxx %}
{% include xxx a=1 b=2 %}
```

The `customTags.filter_include` is run while parsing a template, `context.include()` is run while rendering a template, here is something about `Context.prototype.include`:

```javascript
Context.prototype.include = function (name, localsAst, headerAst, callback) {
  // parameter `name` is the filename
  // parameter `localsAst` is the AST of `with` option
  // parameter `headerAst` is the AST of `a=1 b=2`
};
```

In my option, I recommend you write a custom tag to solve this problem.

**Comment by mercmobily** [comment]
I am a little lost... What do you mean by "The customTags.filter_include is run while parsing a template,"?

Which function shall i use as templare for my custom tag?

**Comment by mercmobily** [comment]
Wait a sec, I think I figured things out.
The implementation of the tag `include` is here:

https://github.com/leizongmin/tinyliquid/blob/07532d946a601c5a7e437e8e973275424e35d338/lib/parser.js#L385

Specifically:

    'include': function (context, name, body) {
      var blocks = arrayRemoveEmptyString(splitText(body, [' ']));
      var filename, withLocals, parameters;
      // support the following pattern:
      // {% include xxx %} or {% include "xxx" %}
      // {% include {{xx}} %} and with filters: {% include {{xx | yy}} %}
      // {% include xxx with yy %}
      // {% include xxx a=1 b=2 %}

But, the "include" tag basically just ends up calling the OpCode INCLUDE, really:

    context.astStack.push(context.astNode(OPCODE.INCLUDE, filename, withLocals, parameters));

In my case, I don't want to use `OPCODE.INCLUDE` since it doesn't do what I want to; instead, I will simply read the file, filter it with my filters, and return the result with something like:

    var ast = tinyliquid.parse(filteredContents);
    context.astStack.push(ast);

Basically, I was confusing the parser's OPCODE for Include and the tag implementation for it.

Right?

Merc.

**Comment by leizongmin** [comment]
You can define a custom tag, the code like this:

```javascript
function filter_include (context, name, body) {
  // generate the new liquid template here, like this:
  var tpl = '{%capture result%}' +
            '{% include _included.md %}' +
            '{%endcapture%}' +
            '{{ result | ryver-markup-markdown }}';
  var ast = tinyliquid.parse(tpl);
  context.astStack.push(ast);
}
```

**Comment by mercmobily** [comment]
Oh my I am so so stupid and you are so so right!!!

**Comment by leizongmin** [comment]
So easy, right ^_^

**Comment by mercmobily** [comment]
If I want to keep all of the original features of the original `include`, could I do something like this?

    function filter_include (context, name, body) {
      // generate the new liquid template here, like this:
      var tpl = '{%capture result%}' +
            '{% include {{body}} %}' +
            '{%endcapture%}' +
            '{{ result | ryver-markup-markdown }}';
      var ast = tinyliquid.parse(tpl);
      context.astStack.push(ast);

...? This way, the extra original features of include will work (like `include xxx with yy`) -- plus I will apply the custom filters. Right?

Thank you for your help!

Merc.

**Comment by leizongmin** [comment]
If your template is:

```
{% filter_include xxx with yyy %}
```

then in this function, the parameters will be:

```javascript
name = 'filter_include';
body = 'xxx with yyy';
```

you want to convert it to the followings:

```
{%capture result%}
  {% include xxx with yyy %}
  {%endcapture%}
{{ result | ryver-markup-markdown }}
```

maybe you can try this:

```javascript
var tpl = '{%capture result%}' +
          '{% include ' + body + ' %}' +
          '{%endcapture%}' +
          '{{ result | ryver-markup-markdown }}';
```

**Comment by mercmobily** [comment]
Sorry, of course, that is what i meant to write! Sorry was typing with my phone...
